src/core_read_parameter_file.c

```diff
@@ -2,217 +2,260 @@
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "SimulationDir");
+    strncpy(ParamTag[NParam], "SimulationDir", MAXTAGLEN);
     ParamAddr[NParam] = run_params->SimulationDir;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "FileWithSnapList");
+    strncpy(ParamTag[NParam], "FileWithSnapList", MAXTAGLEN);
     ParamAddr[NParam] = run_params->FileWithSnapList;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "LastSnapShotNr");
+    strncpy(ParamTag[NParam], "LastSnapShotNr", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->LastSnapShotNr);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "FirstFile");
+    strncpy(ParamTag[NParam], "FirstFile", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->FirstFile);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "LastFile");
+    strncpy(ParamTag[NParam], "LastFile", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->LastFile);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "NumSimulationTreeFiles");
+    strncpy(ParamTag[NParam], "NumSimulationTreeFiles", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->NumSimulationTreeFiles);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "ThreshMajorMerger");
+    strncpy(ParamTag[NParam], "ThreshMajorMerger", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->ThreshMajorMerger);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "RecycleFraction");
+    strncpy(ParamTag[NParam], "RecycleFraction", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->RecycleFraction);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "ReIncorporationFactor");
+    strncpy(ParamTag[NParam], "ReIncorporationFactor", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->ReIncorporationFactor);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "UnitVelocity_in_cm_per_s");
+    strncpy(ParamTag[NParam], "UnitVelocity_in_cm_per_s", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->UnitVelocity_in_cm_per_s);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "UnitLength_in_cm");
+    strncpy(ParamTag[NParam], "UnitLength_in_cm", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->UnitLength_in_cm);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "UnitMass_in_g");
+    strncpy(ParamTag[NParam], "UnitMass_in_g", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->UnitMass_in_g);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "Hubble_h");
+    strncpy(ParamTag[NParam], "Hubble_h", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->Hubble_h);
     ParamID[NParam++] = DOUBLE;
 
-    strcpy(ParamTag[NParam], "ReionizationOn");
+    strncpy(ParamTag[NParam], "ReionizationOn", MAXTAGLEN);
     ParamAddr[NParam] = &(run_params->ReionizationOn);
     ParamID[NParam++] = INT;
 
-    strcpy(ParamTag[NParam], "SupernovaRecipeOn");
+    strncpy(ParamTag[NParam], "SupernovaRecipeOn", MAXTAGLEN);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  Risk is low because the source is a constant string.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)